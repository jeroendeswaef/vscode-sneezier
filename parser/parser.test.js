const antlr4 = require('antlr4');
const Lexer = require('./SvgPathLexer').SvgPathLexer;
const Parser = require('./SvgPathParser').SvgPathParser;
const Visitor = require('./SvgJsonOutputVisitor').SvgJsonOutputVisitor;
// const flatten = require('lodash.flatten');
// const Listener = require('./SvgJsonOutputListener').SvgJsonOutputListener;

const nearley = require("nearley");
const grammar = require("./parser.js");

function flatten(items) {
	const flat = [];
  
	items.forEach(item => {
	  if (Array.isArray(item)) {
		flat.push(...flatten(item));
	  } else {
		flat.push(item);
	  }
	});
  
	return flat;
  }


const getElements = (input) => {
	const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));

	return flatten(
		parser.feed(input).results
		).filter(t => t)
}

describe('Testing svg expressions', () => {
	test('Testing a one-line svg expression', () => {
		const input = "M 60 0 L 120 0 L 180 60 L 180 120 L 120 180 L 60 180 L 0 120 L 0 60";
		expect(getElements(input)).toMatchSnapshot();
	});
	test('Testing a multi-line svg expression', () => {
		const input = "M 60 0 L 120 0 L 180 60 L 180 120 L 120 180 L 60 180\nL 0 120 L 0 60";
		expect(getElements(input)).toMatchSnapshot();
	});
});