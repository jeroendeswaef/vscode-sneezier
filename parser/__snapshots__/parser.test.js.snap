// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing svg expressions Testing a multi-line svg expression 1`] = `
Array [
  Object {
    "type": "helperPoint",
    "x": 60,
    "y": 0,
  },
  Object {
    "type": "lineTo",
    "x": 120,
    "y": 0,
  },
  Object {
    "type": "lineTo",
    "x": 180,
    "y": 60,
  },
  Object {
    "type": "lineTo",
    "x": 180,
    "y": 120,
  },
  Object {
    "type": "lineTo",
    "x": 120,
    "y": 180,
  },
  Object {
    "type": "lineTo",
    "x": 60,
    "y": 180,
  },
  Object {
    "type": "lineTo",
    "x": 0,
    "y": 120,
  },
  Object {
    "type": "lineTo",
    "x": 0,
    "y": 60,
  },
]
`;

exports[`Testing svg expressions Testing a one-line svg expression 1`] = `
Array [
  Object {
    "type": "helperPoint",
    "x": 60,
    "y": 0,
  },
  Object {
    "type": "lineTo",
    "x": 120,
    "y": 0,
  },
  Object {
    "type": "lineTo",
    "x": 180,
    "y": 60,
  },
  Object {
    "type": "lineTo",
    "x": 180,
    "y": 120,
  },
  Object {
    "type": "lineTo",
    "x": 120,
    "y": 180,
  },
  Object {
    "type": "lineTo",
    "x": 60,
    "y": 180,
  },
  Object {
    "type": "lineTo",
    "x": 0,
    "y": 120,
  },
  Object {
    "type": "lineTo",
    "x": 0,
    "y": 60,
  },
]
`;
